### ISSUE TITLE
Внешний вид, позиционирование по БЭМ

### ISSUE THREAD

**Issue by RinatMullayanov** [bem-site/bem-forum-content-ru#1191]
Ознакомился с материалами на https://ru.bem.info/, но не могу для себя прояснить ряд моментов, объясните пожалуйста:
1. Где должен быть описан *внешний вид* **блока** (размер, цвет и т.д.): можно в css-классе блока писать или обязательно в модификаторе блока?
2.  Где должен быть описан *внешний вид* **элемента**: можно в css-классе элемента или обязательно в модификаторе элемента?
3. В методологии указано: *Блок не должен влиять на свое окружение, т. е. блоку не следует задавать внешнюю геометрию (в виде отступов, границ, влияющих на размеры) и позиционирование.* А где тогда указывать свойства влияющие на позиционирование(position, display, float, margin и т.д.)?
4. Все что касается *позиционирования* **элемента** я могу писать в css-классе элементе?

**Comment by tadatuta** [comment]
> 1. Где должен быть описан *внешний вид* **блока** (размер, цвет и т.д.): можно в css-классе блока писать или обязательно в модификаторе блока? Можно ли писать дефолтный внешний вид прямо в в css-классе блока?

Можно смело писать прямо на класс до тех пор, пока не возникнет необходимость отменять такие стили по умолчанию. В этот момент можно вынести их в модификатор.

> 2.  Где должен быть описан *внешний вид* **элемента**: можно в css-классе элемента или обязательно в модификаторе элемента? Можно ли писать дефолтный внешний вид прямо в в css-классе блока?

Аналогично предыдущему ответу.

> 3. В методологии указано: *Блок не должен влиять на свое окружение, т. е. блоку не следует задавать внешнюю геометрию (в виде отступов, границ, влияющих на размеры) и позиционирование.* А где тогда указывать свойства влияющие на позиционирование(position, display, float, margin и т.д.)?

Если дочитать тот же документ до раздела про миксы, там найдется ответ: https://ru.bem.info/methodology/quick-start/#Микс

> 4. Все что касается *позиционирования* **элемента** я могу писать в css-классе элементе?

Да, т.к. элементы — это «внутренняя кухня» блока.

**Comment by RinatMullayanov** [comment]
@tadatuta По поводу 3-го пункта - вы хотите сказать что блок может быть одновременно и элементом родительского блока? И позиционирование описать в css-классе этого элемента?

**Comment by kompolom** [comment]
@RinatMullayanov 
1 Да. можно
2. Здесь играет роль в каком порядке идут стили. При использовании полного БЭМ стека эта проблема решается с помощью зависимостей.

**Comment by tadatuta** [comment]
@RinatMullayanov в целом все так. Для случаев, когда есть вероятность подобных пересечений, нужно просто не экономить на DOM-узлах и делать
```html
<div class="header">
    <div class="header__search-form"> <!-- отвечает за позиционирование -->
        <div class="search-form"></div>
    </div>
</div>
```

**Comment by RinatMullayanov** [comment]
@tadatuta  @kompolom понял, большое спасибо за разъяснения.

**Comment by AlexandrKom** [comment]
т.к. бэм должен быть понятен с первого взгляда, может следующая конструкция более уместна? потому как вряд ли будет понятно, что <div class="header__search-form"> отвечает за позишн.
Может следующая конструкция будет более уместна?

<div class="header header__position">
    <div class="header__search-form">         
    </div>
</div>

**Comment by tadatuta** [comment]
@AlexandrKom `header__search-form` отвечает за позиционирование **формы**, а, скажем, `header__logo` — за позиционирование логотипа. И так далее.

**Comment by AlexandrKom** [comment]
1. Что-то мне подсказывает, что такая обертка не есть тру, т.к. это дополнительный тег, и если каждый блок оборачивать - таких оберток будет довольно много, а ведь можно и обойтись без них. 
2. Вам не кажется, что происходит отрыв информации касающейся бэм-блока (его позишн) от самого блока.
3. если я захочу перенести блок в другой проект/страницу/бэм-блок - мне придется создавать на новом месте новую обертку для него и писать заново css. Не быстрее поправить пару цифр в классе?

**Comment by tadatuta** [comment]
> Что-то мне подсказывает, что такая обертка не есть тру, т.к. это дополнительный тег, и если каждый блок оборачивать - таких оберток будет довольно много, а ведь можно и обойтись без них.

Избавиться от дополнительных оберток можно с помощью [миксов](https://ru.bem.info/methodology/key-concepts/#Микс). Но вообще выбирая между возможностью повторного использования блока целиком VS экономии на обертке достаточно очевидно, что победит повторное использование.

> Вам не кажется, что происходит отрыв информации касающейся бэм-блока (его позишн) от самого блока.

В том-то и дело, что блоки `form`, `logo` или `button` не должны ничего знать про свое позиционирование, т.к. могут быть использованы многократно в совершенно разных местах на странице, с разными отступами и так далее. Про то, что логотип имеет отступ в 22 пикселя от границы шапки должна знать именно шапка (в ней всегда есть такой отступ от края, а логотип где-нибудь подвале вполне может иметь другие отступы и там пусть за них отвечает подвал).

> если я захочу перенести блок в другой проект/страницу/бэм-блок - мне придется создавать на новом месте новую обертку для него и писать заново css. Не быстрее поправить пару цифр в классе?

Вот есть у меня блок `button` и я хочу его использовать 40 раз на проекте: «Войти», «Найти», «Сохранить», «Отмена», «Купить», «Отправить», «Подтвердить» и так далее. Как быть с парой цифр в классе?

**Comment by gexenemy** [comment]
> В методологии указано: Блок не должен влиять на свое окружение, т. е. блоку не следует задавать внешнюю геометрию (в виде отступов, границ, влияющих на размеры) и позиционирование. А где тогда указывать свойства влияющие на позиционирование(position, display, float, margin и т.д.)?
Если дочитать тот же документ до раздела про миксы, там найдется ответ: https://ru.bem.info/methodology/quick-start/#Микс

можете объяснить почему тогда в исходном коде сайта бэма я вижу кучу блоков с позиционированием и без миксов?
![](https://prnt.sc/iv94ch)

**Comment by tadatuta** [comment]
@gexenemy, наверное, можем, но нужные конкретные примеры

**Comment by gexenemy** [comment]
> @gexenemy, наверное, можем, но нужные конкретные примеры

Владимир, вот например блок с именем .nav position:fixed; top: 79; left: 9; либо выше блок .sitemap тоже есть свойства позиционирования и всё задано без миксов, хотя это вроде как противоречит документации?
![](https://cdn1.savepice.ru/uploads/2018/3/23/3cc39fa328754b5c13ea6d619f4ea27e-full.png)

**Comment by vithar** [comment]
У нас на этих выходных хакатон по БЭМ, я буду заниматься сайтом и сделаю ревью реализации блоков. Такие места подправлю.

**Comment by vithar** [comment]
@gexenemy посмотрел на стили блока `.nav`, у него не предполагается переиспользование в других местах и вводить искусственно элемент родителя только для того, чтобы задать ему позиционирование не вижу смысла, это будет тут избыточно.

**Comment by Realetive** [comment]
@vithar какаие-то «двойные стандарты» получаются: [документация](https://ru.bem.info/methodology/css/#%D0%92%D0%BD%D0%B5%D1%88%D0%BD%D1%8F%D1%8F-%D0%B3%D0%B5%D0%BE%D0%BC%D0%B5%D1%82%D1%80%D0%B8%D1%8F-%D0%B8-%D0%BF%D0%BE%D0%B7%D0%B8%D1%86%D0%B8%D0%BE%D0%BD%D0%B8%D1%80%D0%BE%D0%B2%D0%B0%D0%BD%D0%B8%D0%B5) точно определяет, как следует позиционировать блоки. Соответственно, в данном случае `nav` следует сделать элементом «родительского элемента» (или внести уточнение в документ об исключениях).

**Comment by tadatuta** [comment]
БЭМ — это не про жесткие правила, а про рекомендации, которые упрощают жизнь.

Разумеется, в идеале стоит писать все, как написано в доке, но в реальности всегда бывают допущения с учетом знания о внутренней кухне и планах на дальнейшее развитие проекта. Особенно когда разработчик глубоко понимает, что он делает и к чему это может привести.

**Comment by vithar** [comment]
Я отрефакторил, чтобы не было двойных стандартов.

**Comment by lukasvs** [comment]
А почему не стоит в модификаторе блока указывать margin?
Если можно на примере.

**Comment by AlexandrKom** [comment]
@lukasvs 
"В методологии указано: Блок не должен влиять на свое окружение, т. е. блоку не следует задавать внешнюю геометрию (в виде отступов, границ, влияющих на размеры) и позиционирование. А где тогда указывать свойства влияющие на позиционирование(position, display, float, margin и т.д.)?"

А почему не стоит в модификаторе блока указывать margin?
- потому что блок может использоваться неограниченное количество раз на разных страницах, и где то отступы уместны, а где то нет или нужны другие.

как делаем мы, если к блоку на определенной странице или в определенном месте нужен отступ, например margin-right в 10px,  добавляем соответствующий класс class='b-block l-margin__right-normal'
Так же для других отступов и позиционирования

**Comment by vithar** [comment]
Нет, мы миксируем к блоку элемент родителя, который может задавать свои отступы.

`class="logo header__logo"` или `class="button checkout-form__submit"`
